### ISSUE TITLE
A problematic data collection case

### ISSUE THREAD

**Issue by Libbum** [JuliaDynamics/Agents.jl#151]
Here's an example where the current implementation of data collection is causing me some pain.

**Scenario:**
Dynamical system running daily (each `step!`) in the model, agents acting on properties of the model on a yearly basis.

**An attempt at an implementation with data collection:**
```julia
function bream(m)
   m.properties[:lake].bream 
end

function pike(m)
   m.properties[:lake].pike
end

dummy_step(agent, model) = nothing

function run(model, years_to_run=20; 
             agent_properties=[:pos, :oss, :implementation_lag], 
             model_properties=Dict(:model => [bream, pike, m -> m.properties[:lake].nutrients, m -> m.properties[:affectors], m -> m.properties[:tick]]))
    yearly_agent_data = DataFrame()
    daily_model_data = DataFrame()
    step!(model, agent_step!, model_step!, 1)
    for year in 1:years_to_run
        append!(daily_model_data, step!(model, dummy_step, model_step!, 364, model_properties))
        append!(yearly_agent_data, step!(model, agent_step!, model_step!, 1, agent_properties))
    end
    yearly_agent_data, daily_model_data
end
```
I've done a few different things here to highlight some issues
1. No access to current step information (as far as I can tell) for user provided `agent_step!` and `model_step!` functions.
   
Current solution: increment an internal counter in `model_step!`, assigned to `model.properties[:tick]`.

2. There's either an underlying assumption that a model will be run using `n` steps and that's it (which really means `step!` is equivalent to running a complete simulation), or that data is only needed from one call of `step!`.
   
Here, I'm stepping the model through a set number of years, taking yearly data and daily data separately. Since appending to a `DataFrame` is only an option in the internal `_step!` method, I need to use DataFrames `append!` to capture all the data, but now my step values are reset. Of course, the `:tick` values I have are correct, the model and agents have advanced correctly, but I have to now do a bunch of alignment and post processing to get my steps in order inside the resultant `DataFrame`. 

I can of course dump yearly values using a 'complete' simulation
```julia
step!(model, agent_step!, model_step!, 365*years_to_run, agent_properties, when = 365:365:365*years_to_run)
```
but then I have no option to dump daily information from the model. It's not the end of the world if I dropped the `when` clause and saved all of my agent information and model information in the one dataframe for this specific model (100 agents), it's just highly redundant. Scaling this to something larger though..

3. Accessing model data is not streamlined enough, or at least not inline with how the `DataFrame` naming conventions work with the agent properties.

Since there's no option to use array based access to the model (like I've done with `agent_properties` here), we must use the `Dict(:model => [f::Function])` option to obtain model properties via `step!`. In fact, I can't use this method in conjunction with the array based method to get data for each agent along side model data, I can only obtain aggregate `:agent` values. That point at least, is certainly a sane design. It makes no sense to save aggregate agent data in the same `DataFrame` as the individual agent data. Perhaps we can then simplify my case here and ask the question: *how can aggregate and individual agent data be saved to two separate dataframes in the same step?*

Anyhow, as you can see above I've used two named functions which are, in my opinion, unnecessary boilerplate (`bream` and `pike`). Additionally, a few anonymous functions. This is (or at least seems to be) the best practice, since we're outputting this info to a `DataFrame`. There's no point getting our function `f` to return a `Tuple` or `Struct`, and since we're not passing the `DataFrame` itself to this function, we can't return a `DataFrameRow` either. Perhaps there's a way we could use Named Tuples here? Then `push!` could be invoked and something like this would be fine:
```julia
Dict(:model => [m -> (nutrients = m.properties[:lake].nutrients, affectors = m.properties[:affectors], tick = m.properties[:tick])])
```
but for the moment, the current methods just put these values into one column of the `DataFrame` regardless of type.

But I digress.. The multiple function method I use above solves the one column problem, but now there's no control over the naming convention in the `DataFrame`, we loose our symbols and things are replaced with fairly unhelpful titles:
> │ Row  │ #380(model) │ #381(model) │ #382(model) │ bream(model) │ pike(model) │

Thus, more manual processing to align these values to something more manageable.

---

So all in all, I don't yet have a functioning solution that provides me the data I need in a way that doesn't include large amounts of post processing and alignment. Perhaps I've totally overlooked something and this could be reworked in a very simple way (if so, I I'll write up some documentation, or add an example file). Otherwise, I'd like to hear your thoughts on the best way we could facilitate such an outcome.

**Comment by Datseris** [comment]
*(I am assigning Ali because he has done the data collection almost exclusively by himself, while I only contributed in API design. I will of course also comment here)*

**Comment by Datseris** [comment]
@Libbum could you please provide rudimentary implementations for `agent_step!`, the agent type and `model`, so I can run your MWE and also be able to access the datastructures you have?

**Comment by Libbum** [comment]
Sure thing. In meetings all afternoon, but will get that to you later today.

**Comment by kavir1698** [comment]
> No access to current step information (as far as I can tell) for user provided agent_step! and model_step! functions.

> Current solution: increment an internal counter in model_step!, assigned to model.properties[:tick].

Yeah, as you said users can track the steps with a "tick" in properties.

In general, there are two paths in designing the API: include a lot of functionality in the API but at the same time make it more difficult to read the manual and learn the API. Or provide only the most common or most difficult to code functionalities. This approach relies on the user to code some things, but makes the package easier to user. I personally prefer the second approach.

> Since appending to a DataFrame is only an option in the internal _step! method, I need to use DataFrames append! to capture all the data

What if you wrote a function to collect model-level data at correct intervals. Would that solve the need for amending? I have personally gathered individual-level and model-level data into the same DataFrame without a problem.

> how can aggregate and individual agent data be saved to two separate dataframes in the same step?

This was an initial design of the API. It used to produce multiple DataFrames. But the output was not clean and easy to understand. Maybe for special cases, a different function can be introduced, e.g. `step_mixed!`, that does something similar to your `run` function. I would keep the function name separate that `step!` because that is a special use case.

> Perhaps there's a way we could use Named Tuples here?

Would a function that separates columns with tuples into independent columns be a good solution? I once wrote such a function for myself so that data collection would be faster.

**Comment by Libbum** [comment]
Some additional aspects of my model:
```julia
mutable struct Household <: AbstractAgent
  id::Int
  pos::Tuple{Int,Int}
  compliance::Float64 # From willingness to upgrade
  oss::Bool # Is the system upgraded or not?
  information::Bool # Has this agent been told about the upgrade?
  implementation_lag::Int #How long does it take the agent to choose to upgrade?
end

function agent_step!(agent, model) # Daily
    if model.properties[:tick]%365 == 0
        # Once a year households may update their oss
        if !agent.oss && agent.information
            #Agent has an understanding of compliance but has not yet upgraded.
            if rand() < agent.compliance
                agent.oss = true
                if model.properties[:houseowner_type] == Social()
                    # Tell neighbours to increase their compliance
                    neighbor_cells = node_neighbors(agent, model)
                    for neighbor_cell in neighbor_cells
                        node_contents = get_node_contents(neighbor_cell, model)
                        # Skip iteration if the node is empty.
                        length(node_contents) == 0 && continue
                        for neighbourID in node_contents
                            neighbour = id2agent(neighbourID, model)
                            neighbour.compliance = min(neighbour.compliance*1.5, 0.99)
                        end
                    end
                end
            else
                agent.implementation_lag += 1
            end
        end
    end
    return
end

# model.properties is quite large, as is model_step! but here's a selection of things concerning them.
    houseowner_type::HouseOwner = Introverted()
    willingness_to_upgrade::Float64 = 0.2
    threshold_variable::Threshold = Nutrients()
    pike_expectation::Float64 = 1.4 
    target_nutrients::Float64 = 0.7
    nutrient_series::NutrientSeries = Constant()
    municipality::Municipality = Municipality(10, false, false)
    regulate::Bool = true
    lake::LakeState = LakeState(0.7,1.8,20.0)
    tick::Int = 0
    year::Int = 0
    affectors::Int = 0
    outcomes::Outcomes = Outcomes(0,0,0,0,0)

function model_step!(model)
    # Using Parameters.jl
    @unpack year, outcomes, lake = model.properties
    
    model.properties[:tick] += 1
    if model.properties[:tick]%365 == 0
        year += 1
        # Update pollution level: nutrient inflow by affectors
        model.properties[:affectors] = length(filter(a -> !last(a).oss, model.agents))
        # checking the success of upgrade
        if outcomes.year_when_informed > 0 && outcomes.year_of_full_upgrade <= year
            outcomes.upgraded_households_sum = length(filter(a -> last(a).oss, model.agents))
        end
        # Nutrients affect lake dynamics
        update_settings!(model)
        update_nutrients!(model)
    end
    # Update bream stock, pike stock and nutrients (daily)
    model.properties[:lake].bream += dB(lake, model.properties[:lake_parameters])
    model.properties[:lake].pike += dP(lake, model.properties[:lake_parameters])
    model.properties[:lake].nutrients += dNutr(model)

    # Municipality decides on treatment rule
    if model.properties[:tick]%365 == 0 && model.properties[:regulate] #TODO: we may be able to put this above before the daily block?
        # ...
        # A Large bunch of stuff
    end
end
```

That's quickly cut down, so may be missing something, but hopefully you get the drift.

---
> I personally prefer the second approach.

Certainly fine! I think that's probably the best method too. What's needed therefore is a number of recipes in the documentation to help along new comers. This, to me seemed to be 'missing', but now makes complete sense as to why.

 > What if you wrote a function to collect model-level data at correct intervals. Would that solve the need for amending? I have personally gathered individual-level and model-level data into the same DataFrame without a problem.

Highly probable, yeah. Could you give me an example of how you approached this?

> step_mixed!, that does something similar to your run function

Perhaps.. I guess my question here (also connected to the question above), is there a strong reason to not expose `data_collector` methods for users to invoke directly? I'd guess the issue there is the need for global dataframe values and the need (in my case at least) to update them through `model_step!`&mdash;certainly not the cleanest of solutions, so that's fair enough. Which I guess circles back to how you actually wrote your model level collection function.

 > Would a function that separates columns with tuples into independent columns be a good solution?

I'm wondering if we could do some trickery in `data_collecter_aggregate` using `Base.return_types`, which checks the result of the function calls from either the `:agent` or `:model` field_aggregator values. If we get a named tuple, then we should be able to separate `output` and `colnames` accordingly.

I guess there's always the chance that a user wants to keep their tuples inside one column, so perhaps just adding an optional keyword boolean like `separate_data_into_columns=false` in would enable this functionality.

**Comment by kavir1698** [comment]
> Could you give me an example of how you approached this?

It would be something like this:

```julia
function model_collect(model)
  if correct time
    return some value
  else
    return nothing
  end
end

properties = Dict(:model => [model_collect], :agent_field => [aggregate_function])
```

Of course, this doesn't work if you want to collect whole data rather than aggregate ones.

> so perhaps just adding an optional keyword boolean like separate_data_into_columns=false in would enable this functionality.

Yes, this is a quick fix. I would try to implement it. I think it will come handy quite often.

**Comment by Libbum** [comment]
> Of course, this doesn't work if you want to collect whole data rather than aggregate ones.

Yep, that's exactly my roadblock. I want whole data of my model to be output at each step, so these method's don't really allow that (easily) at present. 

I'll try to think of some other possible solutions too.

**Comment by kavir1698** [comment]
> Yep, that's exactly my roadblock. I want whole data of my model to be output at each step, so these method's don't really allow that (easily) at present.

Off the top of my head, I can't think of a DataFrame structure that can accommodate this kind of mixed data.  But this is a fair issue to be solved. I will think about it.

**Comment by Datseris** [comment]
@Libbum for the data collection, the keyword `when` has gotten more functionality.

Will the issue described here be solved, if we allow the keyword `when` to be a function `when(model)`, and if it returns `true` we collect data, but if it is false we do not?

**Comment by Libbum** [comment]
Possibly, I'll take a closer look and see if that makes sense in this case.

**Comment by Datseris** [comment]
Pr #167 implements what I said above, its not yet on master.
